### ISSUE TITLE
Issue with a strict Content Security Policy (CSP)

### ISSUE THREAD

**Issue by dsebastien** [ui-router/visualizer#12]
Hello,

I've tried using the visualizer with my application but noticed the following errors in the console:

```
Refused to load the stylesheet 'blob:http%3A//localhost%3A3000/9ec665e3-b7f8-433f-a4f8-be6d8a0bf92c' because it violates the following Content Security Policy directive: "style-src 'self''".
updateLink @ addStyles.js:242addStyle @ addStyles.js:172addStylesToDom @ addStyles.js:77module.exports @ addStyles.js:37(anonymous function) @ stateVisualizer.css...
addStyles.js:242Refused to load the stylesheet 'blob:http%3A//localhost%3A3000/70a09723-daa2-4135-a979-6e32da4f23f6' because it violates the following Content Security Policy directive: "style-src 'self'".
...
```

I left out some details, but apparently the code tries to add inline styles to the page through the addStylesToDom function (addStylesToDom(newStyles, options);).

The issue with this is that with a Content Security Policy (CSP) active and hardened, using inline styles is forbidden.

I see two solutions to this issue:
- avoid using inline styles entirely (i.e., add/remove classes instead of inline styles)
- provide a way to configure a nounce so that it can be set by your library on the places where you use inline styles and configured in the CSP so that it does not complain about these

I would prefer solution #1 if doable.

Do you think you could solve this?

Thanks for the nice visualizer though :)

**Comment by christopherthielen** [comment]
Sorry for the delay.

I don't really know anything about the CSP issue you describe so I don't know how to solve this.  Part of the problem is that the app needs to support lazy loading.  Its styles are inside the bundled .js and it needs to inject the styling somehow.  I'm using `style-loader` from webpack to do so. 


I'm not sure using classes only would fix the problem because you'd still need the CSS classes injected into your document.  Can you explain further?  

I don't understand the "nounce", how I would publish something like that in the library.

**Comment by dsebastien** [comment]
Hello @christopherthielen !

Thanks for looking at this one. 

The idea when a content security policy is strict is that it can limit security risks by preventing dangerous things like inline scripts and styles, etc. Script tags and including scripts is the most obvious but stylesheets can also be abused. You can find a lot of information about this here: https://developers.google.com/web/fundamentals/security/csp/

If you're using the style loader of webpack then I guess that the following PR should provide a solution to be able to define the nonce on the inserted style tag(s): https://github.com/webpack-contrib/style-loader/pull/155

With that supported by the style loader it'll be possible to define a custom "nonce" attribute with some value to be added to the generate `<link rel="stylesheet" ...` tag(s).

Thus with that you should be able to make the tag look like this in the dom:
```
<link rel="stylesheet" nonce="some_random_string">
```

With that, we'll be able to add "some_random_string" to our CSP to mark that content as trusted and thus not to be blocked by the CSP.

**Comment by christopherthielen** [comment]
Cool. Can you watch that issue, then make a PR (for the visualizer) when that loader PR is merged and released?

**Comment by christophercr** [comment]
Hi, I've created a pull request (#43) to fix this.

As @dsebastien already pointed out, the [PR in style-loader](https://github.com/webpack-contrib/style-loader/pull/155) tpo support custom attributes has been implemented, so it is available since [version 0.14.0](https://github.com/webpack-contrib/style-loader/blob/master/CHANGELOG.md#0140-2017-03-15).
